std/experimental/collections/rcstring.d

```diff
@@ -0,0 +1,833 @@
+    {
+        debug(CollectionRCString)
+        {
+            writefln("rcstring.ctor: begin");
+            scope(exit) writefln("rcstring.ctor: end");
+        }
+        static if (is(Q == immutable) || is(Q == const))
+            _allocator = immutable AllocatorHandler(allocator);
+        else
+            setAllocator(allocator);
+    }
+
+    ///
+    @safe unittest
+    {
+        import std.experimental.allocator : theAllocator, processAllocator;
+
+        auto a = rcstring(theAllocator);
+        auto ca = const rcstring(processAllocator);
+        auto ia = immutable rcstring(processAllocator);
+    }
+
+    /**
+    Constructs a qualified rcstring out of a number of bytes
+    that will use the provided allocator object.
+    For `immutable` objects, a `RCISharedAllocator` must be supplied.
+    If no allocator is passed, the default allocator will be used.
+
+    Params:
+         allocator = a $(REF RCIAllocator, std,experimental,allocator) or
+                     $(REF RCISharedAllocator, std,experimental,allocator)
+                     allocator object
+         bytes = a variable number of bytes, either in the form of a
+                  list or as a built-in rcstring
+
+    Complexity: $(BIGOH m), where `m` is the number of bytes.
+    */
+    this()(ubyte[] bytes...)
+    {
+        this(defaultAllocator!(typeof(this)), bytes);
+    }
+
+    ///
+    @safe unittest
+    {
+        // Create a list from a list of bytes
+        auto a = rcstring('1', '2', '3');
+
+        // Create a list from an array of bytes
+        auto b = rcstring(['1', '2', '3']);
+
+        // Create a const list from a list of bytes
+        auto c = const rcstring('1', '2', '3');
+    }
+
+    /// ditto
+    this(A, this Q)(A allocator, ubyte[] bytes...)
+    if (!is(Q == shared)
+        && (is(A == RCISharedAllocator) || !is(Q == immutable))
+        && (is(A == RCIAllocator) || is(A == RCISharedAllocator)))
+    {
+        this(allocator);
+        _support = typeof(_support)(allocator, bytes);
+    }
+
+    ///
+    @safe unittest
+    {
+        import std.experimental.allocator : theAllocator, processAllocator;
+
+        // Create a list from a list of ints
+        auto a = rcstring(theAllocator, '1', '2', '3');
+
+        // Create a list from an array of ints
+        auto b = rcstring(theAllocator, ['1', '2', '3']);
+    }
+
+    /**
+    Constructs a qualified rcstring out of a string
+    that will use the provided allocator object.
+    For `immutable` objects, a `RCISharedAllocator` must be supplied.
+    If no allocator is passed, the default allocator will be used.
+
+    Params:
+         allocator = a $(REF RCIAllocator, std,experimental,allocator) or
+                     $(REF RCISharedAllocator, std,experimental,allocator)
+                     allocator object
+         s = input string
+
+    Complexity: $(BIGOH m), where `m` is the number of bytes of the input string.
+    */
+    this()(string s)
```

Comment:
Could this be `scope const char[]`? Same question for other constructors and functions below taking `string`, `wstring`, or `dstring`.